_posts/2024-07-19-quarkus-and-leyden.adoc

```diff
@@ -133,20 +106,27 @@ JiT will also lazily upgrade compiled code after it has been executed very many
 
+*** Do we also need to mention deopt and reopt here? I think we need a brief mention ***
+
 Peak optimization is reached when most of the code running is compiled at the highest tier.
 
-Leyden premain addresses the startup problem by caching loaded class info, class linkage, and compilation profile during a training run so it can reuse them to actively populate, link, and compile code in a production run. Leyden is extending CDS to add the extra stuff to the mix.
+Compiling code for peak performance also requires quite some resources, so performing this work ahead of time can also save precious CPU cycles during the application bootstrap, and can manifest in substantial memory savings as well.
+
+But there are some limitations on what we can optimise before runtime just by examining the bytecode. For example, extensive use of reflection prevents the compiler from predicting which symbols will be loaded, linked, and most used at runtime.
+
+The Leyden project has already sucessfully prototyped shifting the work of method compilation from JIT to AOT. Execution and compilation of methods is tracked during the training run. At the end of the run any associated profiling information and compiled code for the method are saved to the AOT archive, allowing them to be quickly mapped back into memory and reused when the application is next run.
+
+As with AOT loading and linking, the training run enables some of the work of profiling and compiling to be done AOT but allows anything not trained still to be compiled via the regular JIT compilation process. Note that method code does not need to have been compiled at the highest tier in order to be saved. Also, when code compiled at a lower tier is restored it can still be recompiled at a higher level. It can also be deoptimized and re-optimized to adapt to different runtime conditions, just as with code compiled in the current runtime. So, the use of AOT compilation is fully integrated into OpenJDK's adpative, dynamic compilation and recompilation model.
 
 == Current status of Project Leyden
 
-There are already experimental https://jdk.java.net/leyden/[early-access builds of Leyden] that can be tested based on https://openjdk.org/jeps/8315737[this draft JEP about Ahead-of-Time Class Linking]. With the https://www.youtube.com/watch?v=lnth19Kf-x0[Leyden Project], the training run idea has been extended to a wider range of data structures, creating the Cache Data Store(CDS). Now the training data contains:
+There are already experimental https://jdk.java.net/leyden/[early-access builds of Leyden] that can be tested based on https://openjdk.org/jeps/8315737[this draft JEP about Ahead-of-Time Class Linking]. With the https://www.youtube.com/watch?v=lnth19Kf-x0[Leyden Project] *** this link is oto out of date -- maybe use the 20204 leyden talk***, the training run idea has been extended to a wider range of data structures embedded in the new AOT cache. Now the training data contains:
```

Comment:
Do we have the link available already?

Reply:
Not yet. The 2024 JVMLS is next week (https://openjdk.org/projects/mlvm/jvmlangsummit/agenda.html) and the Leyden talk will be on Monday afternoon. The talk may not be uploaded straight away though.

Perhaps we might link to the EA release page instead? The EA is advertised at https://jdk.java.net/leyden/

Reply:
n.b. the latest EA (and most notably the release notes hung off that link) does not use the AOTCache terminology introduced in the latest JEP draft. However a new EA is being prepared ready for the JVLMS talk which will use that terminology.